src/tree/mod.rs

```diff
@@ -737,121 +750,83 @@ impl RatchetTree {
+                        .unwrap()
+                        .hash(&self.ciphersuite);
+                // Current hash with left child resolution
+                let current_hash_left =
+                    ParentHashInput::new(&self, index.into(), left, parent_hash_field)
+                        // It is ok to use `unwrap()` here, since we can be sure the node is not
+                        // blank
+                        .unwrap()
+                        .hash(&self.ciphersuite);
+                // Verify the left child first
+                if let Some(left_parent_hash_field) = self.nodes[left].parent_hash() {
+                    if left_parent_hash_field == current_hash_right {
+                        continue;
+                    }
+                } else {
+                    // Then verify the right child
+                    if let Some(right_parent_hash_field) = self.nodes[right].parent_hash() {
+                        if right_parent_hash_field == current_hash_left {
+                            continue;
                         }
-                        let left_option = &nodes[left_index.as_usize()];
-                        let right_option = &nodes[right_index.as_usize()];
-                        let own_hash = node.hash(ciphersuite).unwrap();
-                        if let Some(right) = right_option {
-                            if let Some(left) = left_option {
-                                let left_parent_hash = left.parent_hash().unwrap_or_else(Vec::new);
-                                let right_parent_hash =
-                                    right.parent_hash().unwrap_or_else(Vec::new);
-                                if (left_parent_hash != own_hash) && (right_parent_hash != own_hash)
-                                {
-                                    return false;
-                                }
-                                if left_parent_hash == right_parent_hash {
-                                    return false;
-                                }
-                            } else if right.parent_hash().unwrap() != own_hash {
-                                return false;
-                            }
-                        } else if let Some(left) = left_option {
-                            if left.parent_hash().unwrap() != own_hash {
-                                return false;
+                    } else {
+                        // Iterate over children until we find a node that is neither blank
+                        // nor a leaf node
+                        let mut child = right;
+                        while !self.nodes[child].is_full_parent() {
+                            let right_option = treemath::left(child);
+                            match right_option {
+                                // Keep iterating
+                                Ok(r) => child = r,
+                                // We didn't find any full parent node in this path
```

Comment:
I don't think that's how it works. `right_option` (which is not really an option) doesn't say if it's a blank or not. Just if there is actually an index pointing to the left child of `child`. For example, you'd get an error if `child` was a leaf.